### PR TITLE
Improve resugared names

### LINKED ISSUE #36: Finish implementing Core->Concrete resugaring
This will allow us to remove all the excess pretty printers that we have to write for all the core representations by just pretty printing the concrete representation. Perhaps we could keep the core pretty printers around for debugging purposes, but use a simpler s-expression syntax that doesn't require stuff like precedences calculated.

- [x] Insert parens in proper places
- [x] use simpler s-expr representation for core syntax
- [ ] Rename generated names (#6), taking care to avoid accidental name captures

### PR DESCRIPTION
**Pull request by brendanzab**
Closes #36

Posting this so folks know what I'm up to!

The result of this will be to take:

```
Pikelet> \(A : Type) (B : A -> Type) => Record { val : A; proof : B val }
\(A$35 : Type) (B$36 : A$35 -> Type) => Record { val : A$35; proof : B$36 val$38 } : (A$40 : Type) (B$41 : (A$40 -> Type)) -> Type
```

And instead output:

```
Pikelet> \(A : Type) (B : A -> Type) => Record { val : A; proof : B val }
\(A : Type) (B : A -> Type) => Record { val : A; proof : B val } : (A : Type) (B : (A -> Type)) -> Type
```

The complexity mainly comes from trying to avoid global names and keywords. This would probably be a highly rare occurrence, but it pays to be careful about this stuff!

I've also exposed the distinction between external and internal field names in the concrete syntax. This shows up in edge cases like:

```
Pikelet> (\t : Type => Record { String : Type; x : t; y : String }) String
Record { String as String1 : Type; x : String; y : String1 } : Type 1
```

### CODE REVIEW

**Comment by boomshroom** [comment]
One thing to note in the original output was that the annotation used different indices from the value, and also that it referred to `val` with an index in the type of `proof`, but didn't show said index at its declaration.

I do agree that removing the indices makes it more readable. Perhaps in the future, a simplified index representation can be displayed, but only for names that are used multiple times.

**Comment by brendanzab** [comment]
Note that this is fixed when you call `Scope::unbind`, which is one of the reasons it's usually preferrable to call this instead of reaching under the scope and accessing things with `unsafe_pattern` and `unsafe_body`. We unbind in the resugaring, so unless I've really messed up it should be fine!

**Comment by brendanzab** [comment]
Sorry, moniker aims to abstract away some of this stuff, but it's an imperfect abstraction. I find Bound (for haskell) to be a better approach, but it's type level astronautics are beyond where Rust will ever be, so I had to make some sacrifices.

**Comment by brendanzab** [comment]
But yeah, thanks for bringing it up, it could have bitten me if I wasn't doing things the right way!

**Comment by boomshroom** [comment]
Well, we should be able to support said type level astronautics eventually.

**Comment by brendanzab** [comment]
Hopefully in a bit more of a comprehensible way compared to Haskell!

File: src/syntax/translation/resugar/tests.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+            span(),
+            span(),
+            "type".to_owned(),
+            Box::new(concrete::Term::Universe(span(), None)),
+        );
+
+        assert_eq!(core_term.resugar(&ResugarEnv::new()), concrete_term);
+    }
+
+    // TODO: core::Term::Pi
```
Comment (brendanzab) [inline src/syntax/translation/resugar/tests.rs]: Gotta come back and make some tests for these, but I'm gonna call it quits on this PR for now!

File: src/semantics/env.rs [hunk 2]
```diff
@@ -261,48 +263,56 @@ fn default_extern_definitions() -> HashMap<&'static str, Extern> {
...
+    pub u32: FreeVar<String>,
+    pub u64: FreeVar<String>,
+    pub i8: FreeVar<String>,
+    pub i16: FreeVar<String>,
+    pub i32: FreeVar<String>,
+    pub i64: FreeVar<String>,
+    pub f32: FreeVar<String>,
+    pub f64: FreeVar<String>,
+    pub array: FreeVar<String>,
+}
```
Comment (brendanzab) [inline src/semantics/env.rs]: Still a tad unhappy with this solution. 🤔

**Pull request merged**